### PR TITLE
Draft: Add drivers framework

### PR DESCRIPTION
**Pull request by n1tram1**
Still a few things lacking:

- _Automatic_ registering of the drivers (ie. not having a big match to match all compat strings).
- Properly dissect the _compatible_ string

### CODE REVIEW

File: aarch64_qemuvirt/src/main.rs [hunk 1]
```diff
@@ -18,14 +19,11 @@ extern "C" fn k_main(_device_tree_ptr: usize) -> ! {
     kernel::arch::aarch64::Aarch64::disable_fp_trap();
 
     static PL011: Pl011 = Pl011::new(0x0900_0000);
-    kernel::globals::set_console(&PL011);
+    kernel::globals::set_earlyinit_console(&PL011);
+    // kernel::globals::set_console(&PL011);
```
Comment (Skallwar) [inline aarch64_qemuvirt/src/main.rs:23]:
```suggestion
```

**Review by Skallwar — COMMENTED** [review]
For the allocator maybe we could reuse something like this for the time being: https://docs.rs/acid_alloc/latest/acid_alloc/

File: kernel/src/driver_manager.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    None
+}
+
+fn map_dt_regions(node: &fdt::node::FdtNode) {
+    let pagesize = crate::PagingImpl::get_page_size();
+
+    if let Some(reg) = node.reg() {
+        for memory_region in reg {
+            let start = memory_region.starting_address as usize;
+            let size = memory_region.size.unwrap_or(pagesize);
```
Comment (Skallwar) [inline kernel/src/driver_manager.rs:78]: If there is no size, it should be an error, right?

Comment (n1tram1) [reply kernel/src/driver_manager.rs:78]: Good point, but now I'm just confused about the parser, shouldn't it error out earlier if there is no size ?

Comment (Skallwar) [reply kernel/src/driver_manager.rs:78]: I think we should just panic instead of using a default value

File: kernel/src/driver_manager.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    for compat in compatibles {
+        match compat {
+            "arm,pl011" => return Some(drivers::pl011::Pl011::boxed(console_base)),
+            // "ns16550" => return Some(&drivers::ns16550::Ns16550::new(console_base)),
+            _ => continue,
+        }
+    }
+
+    None
+}
```
Comment (Skallwar) [inline kernel/src/driver_manager.rs:100]: Is this still used with the new MATCHER mechanism?

File: kernel/src/driver_manager.rs (hunk 2 above)
Comment (Skallwar) [reply kernel/src/driver_manager.rs:108]: I think we should just panic instead of using a default value

File: kernel/src/driver_manager.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+//             map_dt_region(&node);
+//             let cons = matching_console_driver(&node);
+//             if cons.is_none() {
+//                 unmap_dt_regions(&node);
+//             }
+//             cons
+//         });
+//
+//     console
+// }
```
Comment (Skallwar) [inline kernel/src/driver_manager.rs:156]: For later?

File: utils/src/init_once.rs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        while self
+            .locked
+            .compare_exchange(
+                false,
+                true,
+                atomic::Ordering::Acquire,
+                atomic::Ordering::Relaxed,
+            )
+            .is_err()
+        {}
```
Comment (Skallwar) [inline utils/src/init_once.rs:31]: I like the idea. Should we create a SpinLock type?

**Review by Skallwar — COMMENTED** [review]
Thanks for the update
- No ns16550 MATCHER?

File: kernel/src/driver_manager.rs (hunk 3 above)
Comment (n1tram1) [reply kernel/src/driver_manager.rs:100]: Thanks for the catch, this is dead code

File: kernel/src/driver_manager.rs (hunk 4 above)
Comment (n1tram1) [reply kernel/src/driver_manager.rs:156]: Dead code as well

File: utils/src/init_once.rs (hunk 5 above)
Comment (n1tram1) [reply utils/src/init_once.rs:31]: In the close future very much yes, but I'd rather keep this as is for now

Comment (Skallwar) [reply utils/src/init_once.rs:31]: Fine by me

File: kernel/src/device_tree.rs [hunk 6]
```diff
@@ -48,3 +54,78 @@ impl DeviceTree {
...
+        .and_then(|node| {
+            map_dt_region(&node);
+            let cons = matching_console_driver(&node);
+            if cons.is_none() {
+                unmap_dt_regions(&node);
+            }
+            cons
+        });
+
+    console
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply kernel/src/device_tree.rs]: done

File: kernel/src/device_tree.rs [hunk 7]
```diff
@@ -48,3 +54,78 @@ impl DeviceTree {
...
+    let console_base = node
+        .reg()?
+        .next()?
+        .starting_address as usize;
+
+    for compat in compatibles {
+        match compat {
+            "arm,pl011" => return Some(Box::new(drivers::pl011::Pl011::new(console_base))),
+            "ns16550" => return Some(Box::new(drivers::ns16550::Ns16550::new(console_base))),
+            _ => continue,
```
[the remark this answers is not in the document]
Comment (n1tram1) [reply kernel/src/device_tree.rs]: workaround with `drivers::CONSOLE_MATCHERS`

**Pull request merged**